### PR TITLE
iox-1394 fix AUTOSAR violations in vector

### PR DESCRIPTION
**Pull request by FerdinandSpitzschnueffler**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [ ] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer

This PR fixes a first set of Axivion warnings in `vector`.

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #1394 (partially)

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -6,7 +6,6 @@
 1. [ ] Branch follows the naming format (`iox-123-this-is-a-branch`)
 1. [ ] Commits messages are according to this [guideline][commit-guidelines]
     - [ ] Commit messages have the issue ID (`iox-#123 commit text`)
-    - [ ] Commit messages are signed (`git commit -s`)
```
Comment (mossmaurice) [inline .github/PULL_REQUEST_TEMPLATE.md:9]: @FerdinandSpitzschnueffler Why did you remove this item? Has the Eclipse policy changed?

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/vector.hpp [hunk 2]
```diff
@@ -154,7 +155,7 @@ class vector
 
     /// @brief returns the capacity of the vector which was given via the template
     ///         argument
-    uint64_t capacity() const noexcept;
+    static constexpr uint64_t capacity() noexcept;
```
Comment (mossmaurice) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/vector.hpp:158]: :heart:

**Review by mossmaurice — COMMENTED** [review]
LGTM in general, two remarks.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 1 above)
Comment (FerdinandSpitzschnueffler) [reply .github/PULL_REQUEST_TEMPLATE.md:9]: Yes, please see [this comment](https://github.com/eclipse-iceoryx/iceoryx/pull/1737#issuecomment-1277466399).

**Review by mossmaurice — APPROVED** [review]

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/vector.hpp [hunk 3]
```diff
@@ -220,21 +221,29 @@ class vector
...
     uint64_t m_size{0U};
 };
-} // namespace cxx
-} // namespace iox
 
+// AXIVION Next Construct AutosarC++19_03-A13.5.5 : intentional implementation with different parameters to enable
+// comparison of vectors with different capacity
 template <typename T, uint64_t CapacityLeft, uint64_t CapacityRight>
-bool operator==(const iox::cxx::vector<T, CapacityLeft>& lhs, const iox::cxx::vector<T, CapacityRight>& rhs) noexcept;
+bool operator==(const vector<T, CapacityLeft>& lhs, const vector<T, CapacityRight>& rhs) noexcept;
```
Comment (elfenpiff) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/vector.hpp:239]: Those operators could also be constexpr. If you have to adjust something in this PR you could add it ... otherwise you can do this when you visit the vector the next time.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/vector.inl [hunk 4]
```diff
@@ -237,16 +244,20 @@ inline bool vector<T, Capacity>::resize(const uint64_t count, const Targs&... ar
     {
         while (count != m_size)
         {
-            emplace_back(args...);
+            IOX_DISCARD_RESULT(emplace_back(args...));
         }
     }
+    else
```
Comment (elfenpiff) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/vector.inl:250]: I think we can rewrite it so that we do not need to use this empty `else`.

```cxx
    if (count < m_size)
    {
        clearFrom(count);
    }
    else
    {
        while(count != m_size)
        {
            IOX_DISCARD_RESULT(emplace_back(args...));
        }
    }

```

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/vector.inl [hunk 5]
```diff
@@ -237,16 +244,20 @@ inline bool vector<T, Capacity>::resize(const uint64_t count, const Targs&... ar
     {
         while (count != m_size)
```
Comment (elfenpiff) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/vector.inl:245]: Wait, why does this work?? This is an infinite loop since count is never incremented?

Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/vector.inl:245]: `m_size` is incremented in `emplace_back`.

**Review by mossmaurice — COMMENTED** [review]
@FerdinandSpitzschnueffler There are some conflicts with `master`. Could you perform a rebase?

**Review by mossmaurice — APPROVED** [review]

**Review by elfenpiff — APPROVED** [review]
